Assembly: prescribed motions.

PaddleStroke opened the pull request:
Adds a simulation tool that let you add motions to joints and create animations.
Initial work by @aiksiongkoh
Edited/reviewed by me

prokoudine commented:
It looks like prescribed motion can only be created for joints that are in the parent assemblies. It doesn't not take joints in subassemblies into consideration.

PaddleStroke commented:
Thanks I'll check this.

prokoudine commented:
That and Iit seems that Rack/Pinion et al. are not supported yet

PaddleStroke commented:
@FreeCAD/design-working-group this new feature needs some testers :)

VincidaB commented on src/Mod/Assembly/CommandCreateSimulation.py:
```diff
@@ -0,0 +1,1067 @@
+        self.currentFrm = self.form.frameSlider.value()
+        self.startFrm = 1
+        self.endFrm = self.form.frameSlider.maximum()
+        if self.startFrm <= self.endFrm:
```
I had to comment this to make play forward/backward work.

VincidaB commented:
https://github.com/FreeCAD/FreeCAD/blob/479823a9efab9f8b77283fd1b48496cc90703d3f/src/Mod/Assembly/CommandCreateSimulation.py#L982-L983

I had to remove this to get the playback working. Should the check be `>` ?

VincidaB commented:
The window for adding a joint the simulation behaves weirdly when on the right screen of a double screen. 

https://github.com/user-attachments/assets/2e88a7e9-7af1-47f3-8436-520516ad589c

PaddleStroke commented:
@VincidaB ah that's not ideal. You have dual monitor? The function that finds the position of the window needs to be updated then.

PaddleStroke commented:
@chennes I have just reviewed this PR again, fixed the few issues I have found and removed the remaining dev clutter. This is IMO good to go. (Note I'm not reviewing my own PR, this was originally Aik-Siong's work)

kadet1090 commented on src/Mod/Assembly/CommandCreateSimulation.py:
```diff
@@ -0,0 +1,1019 @@
+        if self.app_obj.Joint is None:
+            return
+
+        label = self.app_obj.Joint[0].Label + "("
+        if self.app_obj.MotionType == "Linear":
+            label = label + translate("Assembly", "Linear")
+        else:
+            label = label + translate("Assembly", "Angular")
+        label = label + ")"
```
Couldn't it be done like:
```
label = f"({translate("Assembly", "Linear" if self.app_obj.MotionType == "Linear" else "Angular")})"
```

?

PaddleStroke replied:
Your formula is missing `self.app_obj.Joint[0].Label +` which would make the line super long so auto format will split this in several lines which imo will end up being less readable.

kadet1090 replied:
What about:
```python
        type = "Linear" if self.app_obj.MotionType == "Linear" else "Angular"

        self.app_obj.Label = "{label} ({type})".format(
            label=self.app_obj.Joint[0].Label,
            type=translate("Assembly", type)
        )
```

Seems more readable and it is easier to understand how the expected format looks like

PaddleStroke replied:
Can translate take a variable? I seem to recall that some translate function can not. But maybe it's one of the cpp

kadet1090 replied:
translate as function should not have a problem with that - but tool that extracts strings to translate can. If this strings are not used anywhere else they should be probably marked for translation somewhere else with no-op translation. Not sure in this case - good call.

PaddleStroke replied:
Ah well I have included your patch now so hopefully it's ok :)

kadet1090 replied:
We'll see! If there will be a problem, I'll add the required QT_TRANSLATE_NOOPs :)

kadet1090 commented on src/Mod/Assembly/CommandCreateSimulation.py:
```diff
@@ -0,0 +1,1019 @@
+        layout.addWidget(formula_edit, 2, 1)
+
+        # Add the help label above the buttons
+        layout.addWidget(self.help_label0, 3, 0, 1, 2)
+        layout.addWidget(self.help_label1, 4, 0, 1, 2)
+        layout.addWidget(self.help_label2, 5, 0, 1, 2)
+        layout.addWidget(self.help_label3, 6, 0, 1, 2)
+        layout.addWidget(self.help_label4, 7, 0, 1, 2)
+        layout.addWidget(self.help_label5, 8, 0, 1, 2)
+        layout.addWidget(self.help_label6, 9, 0, 1, 2)
+        layout.addWidget(self.help_label7, 10, 0, 1, 2)
```
Can we give better names to those fields?

PaddleStroke replied:
I don't know, I feel there is more useful information in the numbers than there would be in the names

kadet1090 replied:
Fine, looking at code below I think that every naming scheme would have pros and cons :sweat_smile:. Feel free to resolve, I'm not sure how to do that on GH as reviewer...

kadet1090 commented on src/Mod/Assembly/CommandCreateSimulation.py:
```diff
@@ -0,0 +1,1019 @@
+
+        # Slot to toggle help visibility and button text
+        def toggle_help():
+            if self.help_label1.isVisible():
+                self.help_label0.setVisible(False)
+                self.help_label1.setVisible(False)
+                self.help_label2.setVisible(False)
+                self.help_label3.setVisible(False)
+                self.help_label4.setVisible(False)
+                self.help_label5.setVisible(False)
+                self.help_label6.setVisible(False)
+                self.help_label7.setVisible(False)
+                self.help_button.setText(translate("Assembly", "Help"))
+            else:
+                self.help_label0.setVisible(True)
+                self.help_label1.setVisible(True)
+                self.help_label2.setVisible(True)
+                self.help_label3.setVisible(True)
+                self.help_label4.setVisible(True)
+                self.help_label5.setVisible(True)
+                self.help_label6.setVisible(True)
+                self.help_label7.setVisible(True)
+                self.help_button.setText(translate("Assembly", "Hide help"))
+            self.positionDialog()
```
```suggestion
            toggled = not self.help_label0.isVisible()

            self.help_label0.setVisible(toggled)
            self.help_label1.setVisible(toggled)
            self.help_label2.setVisible(toggled)
            self.help_label3.setVisible(toggled)
            self.help_label4.setVisible(toggled)
            self.help_label5.setVisible(toggled)
            self.help_label6.setVisible(toggled)
            self.help_label7.setVisible(toggled)
            
            if toggled:
                self.help_button.setText(translate("Assembly", "Help"))
            else:
                self.help_button.setText(translate("Assembly", "Hide help"))
                
            self.positionDialog()
```

But please test it - I wrote int on github comment.

kadet1090 commented on src/Mod/Assembly/Gui/ViewProviderSimulationGroup.h:
```diff
@@ -0,0 +1,67 @@
+        return false;
+    };
+
+    // protected:
+    /// get called by the container whenever a property has been changed
+    // void onChanged(const App::Property* prop) override;
```
Is this needed?

PaddleStroke replied:
no

kadet1090 commented on src/Mod/Assembly/UtilsAssembly.py:
```diff
@@ -560,7 +600,7 @@ def getBomGroup(assembly):
 
 
-def getJointGroup(assembly):
+def getJointGroup(assembly, create=True):
```
Is this argument used? Also this argument seems like another function.

PaddleStroke replied:
Yes this is fixing a bug where a joint group would be created on the assembly deletion.
Yes creation could be a separate function but I wanted to ensure the existence on joint group when trying to access it.

kadet1090 replied:
Ah, ok so this is failsafe to create if needed. It is OK then. However, the argument is not used in the function?

![image](https://github.com/user-attachments/assets/ab2cc575-ef7d-4b7c-9a7c-a2a27e7c0c2d)

PaddleStroke replied:
Ah well this is a very good point 😄

PaddleStroke replied:
Indeed now I found that I fixed the issue by another mean. And this is just a leftover of my first attempt of fixing [this bug](https://github.com/FreeCAD/FreeCAD/pull/17670).
So I will remove this now

kadet1090 reviewed (changes requested):
Only minor remarks that should be easy to fix

kadet1090 commented:
Also, the https://github.com/FreeCAD/FreeCAD/pull/18020/files#diff-eb837155dc4be0669bbc319c9d9fca3ee55608ecbcca41370df50f9e3741ceaeR176-R180 PR introduces cool helper that can simplify 
```
auto* prop = dynamic_cast<App::PropertyLinkList*>(sim->getPropertyByName("Group"));?
```
into
```
auto prop = sim->getPropertyByName<App::PropertyLinkList>("Group");
```

Would be great to use it - but this can be introduced in another PR that changes this en mass.

PaddleStroke commented:
Added the icons that has somehow disappeared from the PR. Thanks @VincidaB for the icons

Reqrefusion commented:
@maxwxyz conflicts solve

luzpaz commented:
Congrats! Anyone working on the documentation for this ?

galou commented:
Isn't there any issue with the version of the Ondsel solver. I have `2e3659c4bce3e6885269e0cb3d640261b2a91108 src/3rdParty/OndselSolver` that cannot work because the `MbD::LimitIJ::fillqsudot` is not implemented:
```cpp
void MbD::LimitIJ::fillqsudot(FColDsptr col)
{
    (void) col;
	assert(false);
}
```

When I click on `Generate`, I get `FreeCAD: /home/gael/04-sources/freecad/FreeCAD/src/3rdParty/OndselSolver/OndselSolver/LimitIJ.cpp:112: virtual void MbD::LimitIJ::fillqsudot(MbD::FColDsptr): Assertion ``false' failed.`

galou commented:
From, compiled from source.

Both the submodule version from FreeCAD main (2e3659c) and the latest Ondsel main (07785b7) have the same symptoms.

I shouldn't need to manually switch to a different submodule, this would mean that there is an issue (like to forget to commit the submodule version change).

oursland commented:
CI built the PR without issue.  It is likely you have not updated the submodules.

The correct procedure for doing so is to run the following command from within the FreeCAD source tree:

`git submodule update --init --recursive`

galou commented:
It compiles without issue since this is a runtime issue and `git submodule update --init --recursive` brings me back to 2e3659c.

What commit of OndselSolver do you have?
